reposerver/repository/repository.go

```diff
@@ -130,8 +138,38 @@ func (s *Service) ListApps(ctx context.Context, q *apiclient.ListAppsRequest) (*
 }
 
 type operationSettings struct {
-	sem     *semaphore.Weighted
-	noCache bool
+	sem             *semaphore.Weighted
+	noCache         bool
+	allowConcurrent func(root string) bool
+}
+
+// allowConcurrent returns true if given application source can be processed concurrently
+func allowConcurrent(source *v1alpha1.ApplicationSource) func(root string) bool {
+	return func(root string) bool {
+		switch {
+		// Kustomize with parameters requires changing kustomization.yaml file
+		case source.Kustomize != nil:
+			return len(source.Kustomize.Images) == 0 &&
+				len(source.Kustomize.CommonLabels) == 0 &&
+				source.Kustomize.NamePrefix == "" &&
+				source.Kustomize.NameSuffix == ""
+		// Kustomize with parameters requires changing params.libsonnet file
+		case source.Ksonnet != nil:
+			return len(source.Ksonnet.Parameters) == 0
+		// Plugin might change anything so unsafe to process concurrently
+		case source.Plugin != nil:
+			return false
+		}
+
+		appPath := path.Join(root, source.Path)
+		appType, err := discovery.AppType(appPath)
+		if err != nil {
+			return false
+		}
+
+		// Helm based apps might downloads dependencies and cannot be process concurrently
+		return v1alpha1.ApplicationSourceType(appType) != v1alpha1.ApplicationSourceTypeHelm
```

Comment:
Unfortunately helm dependency build fails if run concurrently. So helm based apps from the same revision cannot be processed in parallel. But we can process such apps in different directories will work on it in a separate PR

Reply:
Figured that we can simply add a lock that prevents running concurrent "helm dependency build" in the same directory.